packages/block-editor/src/components/provider/index.js

```diff
@@ -41,86 +24,38 @@ class BlockEditorProvider extends Component {
 			updateEditorSettings,
 			value,
 			resetBlocks,
-			registry,
+			blocks,
+			isLastBlockChangePersistent,
+			onChange,
+			onInput,
 		} = this.props;
 
 		if ( settings !== prevProps.settings ) {
 			updateEditorSettings( settings );
 		}
 
-		if ( registry !== prevProps.registry ) {
-			this.attachChangeObserver( registry );
-		}
-
-		if ( this.isSyncingOutcomingValue ) {
-			this.isSyncingOutcomingValue = false;
-		} else if ( value !== prevProps.value ) {
-			this.isSyncingIncomingValue = true;
-			resetBlocks( value );
-		}
-	}
-
-	componentWillUnmount() {
-		if ( this.unsubscribe ) {
-			this.unsubscribe();
-		}
-	}
-
-	/**
-	 * Given a registry object, overrides the default dispatch behavior for the
-	 * `core/block-editor` store to interpret a state change and decide whether
-	 * we should call `onChange` or `onInput` depending on whether the change
-	 * is persistent or not.
-	 *
-	 * This needs to be done synchronously after state changes (instead of using
-	 * `componentDidUpdate`) in order to avoid batching these changes.
-	 *
-	 * @param {WPDataRegistry} registry     Registry from which block editor
-	 *                                      dispatch is to be overriden.
-	 */
-	attachChangeObserver( registry ) {
-		if ( this.unsubscribe ) {
-			this.unsubscribe();
-		}
-
-		const {
-			getBlocks,
-			isLastBlockChangePersistent,
-		} = registry.select( 'core/block-editor' );
-
-		let blocks = getBlocks();
-		let isPersistent = isLastBlockChangePersistent();
-
-		this.unsubscribe = registry.subscribe( () => {
-			const {
-				onChange,
-				onInput,
-			} = this.props;
-			const newBlocks = getBlocks();
-			const newIsPersistent = isLastBlockChangePersistent();
-			if ( newBlocks !== blocks && this.isSyncingIncomingValue ) {
-				this.isSyncingIncomingValue = false;
-				blocks = newBlocks;
-				isPersistent = newIsPersistent;
-				return;
+		if ( ! this.syncedValue ) {
+			this.syncedValue = blocks;
+		} else if (
+			blocks !== prevProps.blocks ||
+			isLastBlockChangePersistent !== prevProps.isLastBlockChangePersistent
+		) {
+			const hasSyncedValue = this.syncedValue === blocks;
+			if ( ! hasSyncedValue ) {
+				this.syncedValue = blocks;
+				console.log( 'input' );
+				onInput( blocks );
```

Comment:
I'm trying to understand how this is supposed to work but I don't understand yet. In my testing in the previous PR, if you call `onInput` or `onChange` in `componentDidUpdate`, it will only be called once with the last value if you call them synchronously in RichText. which breaks "backticks undo" for instance (curious to know the result of this particular e2e test in this PR)